libebpfdiscoveryskel/src/SyscallProbes.h

```diff
@@ -210,12 +219,100 @@ __attribute__((always_inline)) inline static int handleSysRecvExit(struct pt_reg
 	return handleSysReadExit(ctx, bytesCount);
 }
 
+__attribute__((always_inline)) inline static int handleSysRecvmsgEntry(struct pt_regs* ctx, int fd, struct user_msghdr* msg, int flags) {
+	if (flags & MSG_PEEK) {
+		return 0;
+	}
+
+	if (flags & MSG_TRUNC || flags & MSG_OOB) {
+		// We drop handling the session when these flags are used
+		handleSysCloseEntry(ctx, fd);
+		return 0;
+	}
+
+	if (msg == NULL) {
+		return 0;
+	}
+
+	struct DiscoveryGlobalState* globalStatePtr = getGlobalState();
+	if (globalStatePtr == NULL || globalStatePtr->isCollectingDisabled) {
+		return 0;
+	}
+
+	struct DiscoveryAllSessionState* allSessionStatePtr = getAllSessionState();
+	if (allSessionStatePtr == NULL) {
+		return 0;
+	};
+
+	__u64 pidTgid = bpf_get_current_pid_tgid();
+
+	struct DiscoveryTrackedSessionKey trackedSessionKey = {};
+	trackedSessionKey.pid = pidTgidToPid(pidTgid);
+	trackedSessionKey.fd = fd;
+
+	if (bpf_map_lookup_elem(&trackedSessionsMap, &trackedSessionKey) == NULL) {
+		// If the read call is not part of a being handled session, stop handling the syscall
+		return 0;
+	}
+
+	struct ReadIovArgs readIovArgs = {};
+	readIovArgs.fd = trackedSessionKey.fd;
+	bpf_probe_read(&readIovArgs.iov, sizeof(struct iovec*), &msg->msg_iov);
+	bpf_probe_read(&readIovArgs.iovlen, sizeof(size_t), &msg->msg_iovlen);
+
+	bpf_map_update_elem(&runningReadIovArgsMap, &pidTgid, &readIovArgs, BPF_ANY);
+	
+	return 0;
+}
+
+__attribute__((always_inline)) inline static int handleSysRecvmsgExit(struct pt_regs* ctx, ssize_t bytesCount) {
+	struct DiscoveryGlobalState* globalStatePtr = getGlobalState();
+	if (globalStatePtr == NULL || globalStatePtr->isCollectingDisabled) {
+		return 0;
+	}
+
+	struct DiscoveryAllSessionState* allSessionStatePtr = getAllSessionState();
+	if (allSessionStatePtr == NULL) {
+		return 0;
+	};
+
+	__u64 pidTgid = bpf_get_current_pid_tgid();
+
+	struct ReadIovArgs* readIovArgsPtr = (struct ReadIovArgs*)bpf_map_lookup_elem(&runningReadIovArgsMap, &pidTgid);
+	if (readIovArgsPtr == NULL) {
+		return 0;
+	}
+
+	if (readIovArgsPtr->iov == NULL) {
+		return 0;
+	}
+
+	for (size_t i = 0; i < PROTOCOL_VEC_LIMIT && i < readIovArgsPtr->iovlen; i++) {
+		struct ReadBufArgs readBufArgs = {
+				.fd = readIovArgsPtr->fd,
+		};
+
+		bpf_probe_read(&readBufArgs.buf, sizeof(void*), &readIovArgsPtr->iov[i].iov_base);
+		if (readBufArgs.buf == NULL) {
+			break;
```

Comment:
NO_MORE_DATA event should be sent to userspace.

Reply:
I've implemented it, but can't we just call `deleteTrackedSession` there?

Reply:
No, we have to inform the userspace so that data from there is cleaned-up too.